File: aoc15/ocaml/day_23/day_23.ml
type inst = { op : string; reg : char; offset : int };;

let parse line =
  let split = String.split_on_char ' ' line in
  let op = List.hd split in
  let reg = (List.nth split 1).[0] in
  let offset =
    if op.[0] = 'j' then int_of_string (List.nth split (List.length split - 1))
    else 0
  in
  { op; reg; offset }
in
let insts = Input.parse_lines parse "input" in

let exec init_a =
  let ip = ref 0 in
  let a = ref init_a in
  let b = ref 0 in
  while !ip >= 0 && !ip < List.length insts do
    let inst = List.nth insts !ip in
    ip := !ip + 1;
    let reg = if inst.reg = 'a' then a else b in
    match inst.op with
    | "hlf" -> reg := !reg / 2
    | "tpl" -> reg := !reg * 3
    | "inc" -> reg := !reg + 1
    | "jmp" -> ip := !ip + inst.offset - 1
    | "jie" -> if !reg mod 2 = 0 then ip := !ip + inst.offset - 1
    | "jio" -> if !reg = 1 then ip := !ip + inst.offset - 1
    | _ -> ()
  done;
  !b
in

let first = exec 0 in
Printf.printf "First: %d\n" first;
let second = exec 1 in
Printf.printf "Second: %d\n" second

File: aoc15/go/day_23/day_23.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Inst struct {
	Opcode string
	Reg    byte
	Offset int
}

func parse(line string) Inst {
	split := strings.Fields(line)
	switch split[0] {
	case "hlf", "tpl", "inc":
		return Inst{Opcode: split[0], Reg: split[1][0]}
	case "jmp":
		offset, _ := strconv.Atoi(split[1])
		return Inst{Opcode: split[0], Offset: offset}
	case "jie", "jio":
		offset, _ := strconv.Atoi(split[2])
		return Inst{Opcode: split[0], Reg: split[1][0], Offset: offset}
	}
	return Inst{}
}

func exec(insts []Inst, init int) int {
	regs := map[byte]int{'a': init, 'b': 0}
	ip := 0
	for ip >= 0 && ip < len(insts) {
		i := insts[ip]
		ip++
		switch i.Opcode {
		case "hlf":
			regs[i.Reg] /= 2
		case "tpl":
			regs[i.Reg] *= 3
		case "inc":
			regs[i.Reg] += 1
		case "jmp":
			ip += i.Offset - 1
		case "jie":
			if regs[i.Reg]%2 == 0 {
				ip += i.Offset - 1
			}
		case "jio":
			if regs[i.Reg] == 1 {
				ip += i.Offset - 1
			}
		}
	}
	return regs['b']
}

func main() {
	input, _ := os.Open("input")
	scanner := bufio.NewScanner(input)

	insts := []Inst{}
	for scanner.Scan() {
		insts = append(insts, parse(scanner.Text()))
	}

	fmt.Println("First:", exec(insts, 0))
	fmt.Println("Second:", exec(insts, 1))
}
